File: app/assets/javascripts/dashboard/ContestService.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Contest Service
     * @author lawrence.daniels@gmail.com
     */
    app.factory('ContestService', function ($http, $log, $q, toaster, QuoteService) {
        var service = {};

        ///////////////////////////////////////////////////////////////
        //          Contest C.R.U.D.
        ///////////////////////////////////////////////////////////////

        service.createContest = function (form) {
            return $http({method: "PUT", url: "/api/contest", data: angular.toJson(form)});
        };

        service.deleteContest = function (contestId) {
            return $http.delete('/api/contest/' + contestId);
        };

        service.joinContest = function (contestId, playerInfo) {
            return $http({
                method: 'PUT',
                url: "/api/contest/" + contestId + "/player",
                data: angular.toJson(playerInfo)
            });
        };

        service.quitContest = function (contestId, playerId) {
            return $http.delete('/api/contest/' + contestId + "/player/" + playerId);
        };

        service.startContest = function (contestId) {
            return $http.get("/api/contest/" + contestId + "/start");
        };

        ///////////////////////////////////////////////////////////////
        //          Contest Finders
        ///////////////////////////////////////////////////////////////

        service.findContests = function (searchOptions) {
            if(!searchOptions) throw "SearchOptions is null or undefined";
            else {
                return $http({method: 'POST', url: '/api/contests/search', data: angular.toJson(searchOptions)})
                    .then(function (response) {
                        return response.data;
                    });
            }
        };

        service.getContestByID = function (contestId) {
            if(!contestId) throw "getContestByID: Contest ID is null or undefined";
            else return $http.get("/api/contest/" + contestId);
        };

        service.getParticipantByID = function (contestId, playerId) {
            if(!contestId) throw "getParticipantByID: Contest ID is null or undefined";
            else if(!playerId) throw "getParticipantByID: Player ID is null or undefined";
            else return $http.get("/api/contest/" + contestId + "/player/" + playerId);
        };

        service.getCashAvailable = function (contest, playerId) {
            if(contest) {
                var player = service.findPlayerByID(contest, playerId);
                return player ? player.fundsAvailable : 0.00;
            }
            return 0.00;
        };

        service.getRankings = function (contestId) {
            if(!contestId) throw "getRankings: Contest ID is null or undefined";
            else return $http.get("/api/contest/" + contestId + "/rankings");
        };

        service.getContestsByPlayerID = function (playerId) {
            if(!playerId) throw "getContestsByPlayerID: Player ID is null or undefined";
            else return $http.get('/api/contests/player/' + playerId);
        };

        service.getEnrichedOrders = function (contestId, playerId) {
            if(!contestId) throw "getEnrichedOrders: Contest ID is null or undefined";
            else if(!playerId) throw "getEnrichedOrders: Player ID is null or undefined";
            else return $http.get('/api/contest/' + contestId + '/orders/' + playerId);
        };

        service.getEnrichedPositions = function (contestId, playerId) {
            if(!contestId) throw "getEnrichedPositions: Contest ID is null or undefined";
            else if(!playerId) throw "getEnrichedPositions: Player ID is null or undefined";
            else return $http.get('/api/contest/' + contestId + '/positions/' + playerId);
        };

        /////////////////////////////////////////////////////////////////////////////
        //			Participants
        /////////////////////////////////////////////////////////////////////////////

        service.getTotalInvestment = function (playerId) {
            if(!playerId) throw "getTotalInvestment: Player ID is null or undefined";
            else return $http.get('/api/contests/player/' + playerId + '/totalInvestment');
        };

        service.findPlayerByID = function (contest, playerId) {
            var participants = contest ? (contest.participants || []) : [];
            for (var n = 0; n < participants.length; n++) {
                var participant = participants[n];
                if (participant.OID() === playerId) {
                    return participant;
                }
            }
            return null;
        };

        service.findPlayerByName = function (contest, playerName) {
            var participants = contest ? (contest.participants || []) : [];
            for (var n = 0; n < participants.length; n++) {
                var participant = participants[n];
                if (participant.name === playerName) {
                    return participant;
                }
            }
            return null;
        };

        service.getPlayerRankings = function (contest, playerName) {
            if (!contest || !contest.name) return [];
            else {
                // if the rankings have never been loaded ...
                if (contest.rankings === undefined) {
                    contest.rankings = {};
                    $log.info("Loading Contest Rankings for '" + contest.name + "'...");
                    service.getRankings(contest.OID())
                        .success(function (participants) {
                            contest.rankings.participants = participants;
                            if (participants.length) {
                                contest.rankings.leader = participants[0];
                                contest.rankings.player = playerName ? service.findPlayerByName(contest.rankings, playerName) : null;
                            }
                        })
                        .error(function (response) {
                            toaster.pop('error', 'Error!', "Error loading play rankings");
                            $log.error(response.error)
                        });
                }

                // if the rankings were loaded, but the player is not set
                else if(playerName && contest.rankings.player == null) {
                    contest.rankings.player = playerName ? service.findPlayerByName(contest.rankings, playerName) : null;
                }

                return contest.rankings;
            }
        };

        /////////////////////////////////////////////////////////////////////////////
        //			Miscellaneous
        /////////////////////////////////////////////////////////////////////////////

        service.getChart = function (contestId, participantName, chartName) {
            // build the appropriate URL
            var uriString = (chartName == "gains" || chartName == "losses")
                ? "/api/charts/performance/" + chartName + "/" + contestId + "/" + participantName
                : "/api/charts/exposure/" + chartName + "/" + contestId + "/" + participantName;

            // load the chart representing the securities
            return $http.get(uriString).then(function (response) {
                return response.data;
            });
        };

        /////////////////////////////////////////////////////////////////////////////
        //			Chat
        /////////////////////////////////////////////////////////////////////////////

        service.sendChatMessage = function (contestId, message) {
            return $http({
                method: 'PUT',
                url: '/api/contest/' + contestId + '/chat',
                data: angular.toJson(message)
            });
        };

        /////////////////////////////////////////////////////////////////////////////
        //			Positions & Orders
        /////////////////////////////////////////////////////////////////////////////

        service.createOrder = function (contestId, playerId, order) {
            return $http({
                method: "PUT",
                url: "/api/order/" + contestId + "/" + playerId,
                data: angular.toJson(order)
            });
        };

        service.deleteOrder = function (contestId, playerId, orderId) {
            return $http.delete('/api/order/' + contestId + '/' + playerId + '/' + orderId);
        };

        service.getHeldSecurities = function (playerId) {
            return $http.get("/api/positions/" + playerId);
        };

        service.orderQuote = function (symbol) {
            $log.info("Loading symbol " + symbol + "'...");
            return $http.get('/api/quotes/order/symbol/' + symbol)
                .then(function (response) {
                    var quote = response.data;
                    if (quote.symbol) {
                        $log.info("Setting lastSymbol as " + quote.symbol);
                        QuoteService.lastSymbol = quote.symbol;
                    }
                    return quote;
                });
        };

        return service;
    });

})();

File: app-web-angularjs/src/main/scala/com/shocktrade/javascript/dashboard/ContestService.scala
package com.shocktrade.javascript.dashboard

import com.shocktrade.javascript.dashboard.ContestService.{ContestHost, MarginMarketValue}
import com.shocktrade.javascript.forms.PlayerInfoForm
import com.shocktrade.javascript.models.contest._
import org.scalajs.angularjs.http.Http
import org.scalajs.angularjs.toaster.Toaster
import org.scalajs.angularjs.{Service, angular}
import org.scalajs.dom.console
import org.scalajs.nodejs.util.ScalaJsHelper._

import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.scalajs.js.JSConverters._
import scala.scalajs.js.annotation.ScalaJSDefined
import scala.util.{Failure, Success}

/**
  * Contest Service
  * @author lawrence.daniels@gmail.com
  */
class ContestService($http: Http, toaster: Toaster) extends Service {

  ///////////////////////////////////////////////////////////////
  //          Basic C.R.U.D.
  ///////////////////////////////////////////////////////////////

  def deleteContest(contestId: String) = {
    $http.delete[js.Dynamic](s"/api/contest/$contestId")
  }

  def joinContest(contestId: String, playerInfo: PlayerInfoForm) = {
    $http.put[Contest](s"/api/contest/$contestId/player", playerInfo)
  }

  def quitContest(contestId: String, playerId: String) = {
    $http.delete[Contest](s"/api/contest/$contestId/player/$playerId")
  }

  def startContest(contestId: String) = {
    $http.get[Contest](s"/api/contest/$contestId/start")
  }

  def updateContestHost(contestId: String, host: String) = {
    $http.post[js.Dynamic](s"/api/contest/$contestId/host", new ContestHost(host))
  }

  ///////////////////////////////////////////////////////////////
  //          Contest Finders
  ///////////////////////////////////////////////////////////////

  def findContests(searchOptions: ContestSearchOptions) = {
    $http.post[js.Array[Contest]]("/api/contests/search", searchOptions)
  }

  def getContestByID(contestId: String) = {
    $http.get[Contest](s"/api/contest/$contestId")
  }

  def getParticipantByID(contestId: String, playerId: String) = {
    $http.get[js.Dynamic](s"/api/contest/$contestId/player/$playerId")
  }

  def getCashAvailable(contest: Contest, playerId: String) = {
    val participant = contest.participants.flatMap(_.find(_._id.contains(playerId)).orUndefined)
    participant.flatMap(_.cashAccount.flatMap(_.cashFunds)) getOrElse 0.0d
  }

  def getRankings(contestId: String) = {
    $http.get[js.Array[ParticipantRanking]](s"/api/contest/$contestId/rankings")
  }

  def getContestsByPlayerID(playerId: String) = {
    $http.get[js.Array[Contest]](s"/api/contests/player/$playerId")
  }

  def getEnrichedOrders(contestId: String, playerId: String) = {
    $http.get[js.Array[Order]](s"/api/contest/$contestId/orders/$playerId")
  }

  def getEnrichedPositions(contestId: String, playerId: String) = {
    $http.get[js.Array[Position]](s"/api/contest/$contestId/positions/$playerId")
  }

  /////////////////////////////////////////////////////////////////////////////
  //			Participants
  /////////////////////////////////////////////////////////////////////////////

  def getMarginMarketValue(contestId: String, playerId: String) = {
    $http.get[MarginMarketValue](s"/api/contest/$contestId/margin/$playerId/marketValue")
  }

  def getPlayerRankings(contest: Contest, playerID: String): js.UndefOr[Rankings] = {
    if (isDefined(contest) && isDefined(contest.name)) {
      // if the rankings have never been loaded ...
      if (!isDefined(contest.rankings)) {
        // create the rankings object: { participants: [], leader:null, player:null }
        contest.rankings = new Rankings()

        console.log(s"Loading Contest Rankings for '${contest.name}'...")
        for {
          contestId <- contest._id
          userID = playerID
        } {
          getRankings(contestId) onComplete {
            case Success(participantRankings) =>
              console.log(s"participantRankings = ${angular.toJson(participantRankings)}")
              contest.rankings.foreach { rankings =>
                rankings.participants = participantRankings
                rankings.leader = participantRankings.headOption.orUndefined
                rankings.player = participantRankings.find(p => p._id.contains(userID) || p.facebookID.contains(userID)).orUndefined
              }
            case Failure(e) =>
              toaster.error("Error loading player rankings")
              console.error(s"Error loading player rankings: ${e.getMessage}")
              e.printStackTrace()
          }
        }
      }

      // if the rankings were loaded, but the player is not set
      else if (contest.rankings.nonEmpty && contest.rankings.exists(_.player.isEmpty)) {
        for {
          rankings <- contest.rankings
          userID = playerID
        } {
          rankings.player = rankings.participants.find(p => p._id.contains(userID) || p.facebookID.contains(userID)).orUndefined
        }
      }

      contest.rankings
    }

    else New[Rankings]
  }

  def getTotalInvestment(playerID: String) = {
    $http.get[js.Dynamic](s"/api/contests/player/$playerID/totalInvestment")
  }

  /////////////////////////////////////////////////////////////////////////////
  //			Charts
  /////////////////////////////////////////////////////////////////////////////

  def getExposureChartData(contestID: String, playerID: String, exposure: String) = {
    $http.get[js.Array[js.Object]](s"/api/charts/exposure/$exposure/$contestID/$playerID")
  }

  def getChart(contestID: String, playerName: String, chartName: String) = {
    // determine the chart type
    val chartType = if (chartName == "gains" || chartName == "losses") "performance" else "exposure"

    // load the chart representing the securities
    $http.get[js.Dynamic](s"/api/charts/$chartType/$chartName/$contestID/$playerName")
  }

  /////////////////////////////////////////////////////////////////////////////
  //			Chat
  /////////////////////////////////////////////////////////////////////////////

  def sendChatMessage(contestId: String, message: Message) = {
    $http.put[js.Array[Message]](s"/api/contest/$contestId/chat", message)
  }

  /////////////////////////////////////////////////////////////////////////////
  //			Positions & Orders
  /////////////////////////////////////////////////////////////////////////////

  def deleteOrder(contestId: String, playerId: String, orderId: String) = {
    $http.delete[Contest](s"/api/order/$contestId/$playerId/$orderId")
  }

  def getHeldSecurities(playerId: String) = {
    $http.get[js.Array[String]](s"/api/positions/$playerId")
  }

}

/**
  * Contest Service Companion
  * @author lawrence.daniels@gmail.com
  */
object ContestService {

  /**
    * Contest Host
    * @param host the given host name
    */
  @ScalaJSDefined
  class ContestHost(val host: String) extends js.Object

  /**
    * Margin Market Value
    * @author lawrence.daniels@gmail.com
    */
  @js.native
  trait MarginMarketValue extends js.Object {
    var _id: js.UndefOr[String] = js.native
    var name: String = js.native
    var marginMarketValue: Double = js.native
  }

}